libtextsecure/message_receiver.js

```diff
@@ -154,42 +126,49 @@ MessageReceiver.prototype.extend({
     stringToArrayBuffer: function(string) {
         return new dcodeIO.ByteBuffer.wrap(string, 'binary').toArrayBuffer();
     },
-    addToCache: function(envelope, plaintext) {
-        console.log('addToCache', this.getEnvelopeId(envelope));
-        return new Promise(function(resolve, reject) {
-            var id = this.getEnvelopeId(envelope);
-            var string = this.arrayBufferToString(plaintext);
-            var unprocessed = new Whisper.Unprocessed({
-                id: id,
-                envelope: string,
-                timestamp: Date.now(),
-                attempts: 1
-            });
-            return unprocessed.save().then(resolve, reject);
+    getAllFromCache: function() {
+        console.log('getAllFromCache');
+        return textsecure.storage.unprocessed.getAll().then(function(items) {
+            console.log('getAllFromCache loaded', items.length, 'saved envelopes');
+
+            _.forEach(items, function(item) {
+                var attempts = 1 + (item.attempts || 0);
+                if (attempts >= 5) {
+                    console.log('getAllFromCache final attempt for envelope', item.id);
+                    textsecure.storage.unprocessed.remove(item.id);
+                } else {
+                    textsecure.storage.unprocessed.update(item.id, {attempts: attempts});
+                }
+            }.bind(this));
+
+            return items;
```

Comment:
Maybe we removed some items from storage or updated them, but we still return them without blocking on the write to storage. Is it safe? Might we cause a race between remove/update and an update after decryption? Possibly the queuing in queueCache ameliorates that danger, so just something to be mindful of.